extensions/amp-analytics/0.1/amp-analytics.js

```diff
@@ -0,0 +1,232 @@
+import {urlReplacementsFor} from '../../../src/url-replacements';
+
+import {addListener} from './instrumentation';
+import {AnalyticsConfig} from './vendors';
+
+export class AmpAnalytics extends AMP.BaseElement {
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return true;
+  }
+
+  /** @override */
+  buildCallback() {
+
+    /**
+     * @private {?JSONValue} The analytics config associated with the tag
+     */
+    this.config_ = null;
+
+    /**
+     * @private {?string} Predefinedtype associated with the tag. If specified,
+     * the config from the predefined type is merged with the inline config
+     */
+    this.type_ = null;
+
+    /**
+     * @private {Object.<string, string>} A map of request names to the request
+     * format string used by the tag to send data
+     */
+    this.requests_ = {};
+    this.element.setAttribute('aria-hidden', 'true');
+
+    this.mergeConfigs_();
+
+    if (this.hasOptedOut_()) {
+      // Nothing to do when the user has opted out.
+      log.fine(this.getName_(), "User has opted out. No hits will be sent.");
+      return;
+    }
+
+    this.generateRequests_();
+
+    if (!Array.isArray(this.config_['triggers'])) {
+      log.error(this.getName_(), 'No triggers were found in the config. No ' +
+          'analytics data will be sent.');
+      return;
+    }
+
+    // Trigger callback can be synchronous. Do the registration at the end.
+    for (let k = 0; k < this.config_['triggers'].length; k++) {
+      const trigger = this.config_['triggers'][k];
+      if (!trigger['on'] || !trigger['request']) {
+        log.warn(this.getName_(), '"on" and "request" attributes are ' +
+            'required for data to be collected.');
+        continue;
+      }
+      addListener(this.getWin(), trigger['on'],
+          this.handleEvent_.bind(this, trigger));
+    }
+  }
+
+  /**
+   * Merges various sources of configs and stores them in a member variable.
+   *
+   * Order of precedence for configs from highest to lowest:
+   * - Remote config: specified through an attribute of the tag.
+   * - Inline config: specified insize the tag.
+   * - Predefined config: Defined as part of the platform.
+   *
+   * @private
+   */
+  mergeConfigs_() {
+    // TODO(btownsend, #871): Implement support for remote configuration.
+    const remote = {};
+
+    const inline = JSON.parse(this.element.textContent) || {};
+    const config = AnalyticsConfig[this.element.getAttribute('type')] || {};
+    mergeObjects(remote, mergeObjects(inline, config));
+    this.config_ = config;
+  }
+
+  /**
+   * @return {boolean} true if the user has opted out.
+   */
+  hasOptedOut_() {
+    if (!this.config_['optout']) {
+      return false;
+    }
+
+    const props = this.config_['optout'].split('.');
+    let obj = this.getWin();
+    for (let i = 0; i < props.length; i++) {
+      if (!obj) {
+        return false;
+      }
+      obj = obj[props[i]];
+    }
+    return obj();
```

Comment:
reads strange: you are calling an object. Especially since you initialize it above as a window. I know it's correct, but perhaps some naming could be fixed?